ansible/roles/preflights/tasks/main.yaml

```diff
@@ -0,0 +1,393 @@
+      - 179   # calico-node bird (used for BGP)
+    when:
+      - not kubeadm_already_run.stat.exists
+      - "calico_encapsulation != 'vxlan'"
+    changed_when: False
+
+  - name: verify control-plane to control-plane connectivity
+    command: ping -c 2 {{ item }}
+    with_items: "{{ groups['control-plane'] }}"
+    when: "'control-plane' in group_names"
+    changed_when: False
+
+  - name: verify node to control-plane connectivity
+    command: ping -c 2 {{ item }}
+    with_items: "{{ groups['control-plane'] }}"
+    when: "'node' in group_names"
+    changed_when: False
+
+  # Every node should be able to reach all other nodes.
+  # Use a random sampling of nodes to avoid quadratic complexity.
+  - name: verify node to node connectivity with random sample
+    command: ping -c 2 {{ item }}
+    with_random_choice: "{{ groups['node'] | map('extract', hostvars, 'inventory_hostname') | list }}"
+    when: "'node' in group_names"
+    changed_when: False
+
+
+  # Need to properly handle larger clusters
+  # Consider forks - where each task runs in some batch( default 20) and serial where each play is split into batches
+  # Hardcoding a specific node will not work as they may be run in a different Ansible batches
+  # The `ansible_date_time` var is collected during fact gathering
+  # This task assumes that the fact gathering runs around the same time
+  # So adding an extra second of allowance for each batch of forks that will run
+  - name: compare time is within 30 seconds of the first node in this batch
+    fail:
+      msg: "The time difference between this node and the first node in this batch is greater than 30 seconds"
+    failed_when: ((ansible_date_time.epoch | int) - (hostvars[ansible_play_batch[0]]['ansible_date_time']['epoch'] | int)) | abs > ((30 + (ansible_play_batch|length / ansible_forks)) | int)
+
+  - block:
+    - name: check if keepalived is already configured
+      stat:
+        path: "/etc/kubernetes/manifests/keepalived.yaml"
+      delegate_to: "{{ groups['control-plane'][0] }}"
+      register: keepalived_file
+
+    - block:
+      - name: check reachability to keepalived vip
+        command: ping -c 2 {{ keepalived.virtual_ip }}
+        changed_when: False
+      when:
+        - keepalived_file.stat.exists
+
+    - block:
+      - name: check if keepalived vip is already present
+        command: ping -c 2 {{ keepalived.virtual_ip }}
+        register: pingout
+        changed_when: pingout.rc == 0
+        failed_when: pingout.rc == 0
+
+      - block:
+        - block:
+          - name: detect a keepalived interface
+            shell: ip route get "{{ keepalived.virtual_ip }}" | grep -Po '(?<=(dev )).*(?= src| proto)'
+            register: keepalived_intf
+
+          - name: set keepalived interface
+            set_fact:
+               keepalived:
+                 interface: "{{ keepalived_intf.stdout }}"
+            when: keepalived.interface == ""
+
```

Comment:
@dlipovetsky would equivalent checks still be valid for kube-vip? The only input is [interface](https://github.com/mesosphere/cluster-api-provider-preprovisioned/pull/5/files#diff-e0bb8f569e140bdc6127d8dfaa3677d0f444cad8e845dbb1e1e45f6009531e8eR68). The IP would be the same that is passed to CAPP.

Reply:
removing these blocks in the meantime.